### DISCUSSION TITLE
Allow kadalu volumes to be mounted externally (outside k8s)

### DISCUSSION THREAD

**Discussion by amarts** [kadalu/kadalu#563] [Feature Requests]
few users started asking for such features.

A storage volume started with kadalu operator. Now, from an external node which has access to reach gluster server pods wants to mount the volume.

There can be 2 type of users, 1 - use complete volume, or 2nd type is those who wants to use only 1 PV out of the storage on external node.

**Comment by amarts** [comment]
Possible Solution:

Phase1 -  Static

1. Build a very simple REST API server app which serves gluster volfile as response when a particular GET call is made.
2. In this server - keep a watch on configMap, and like how we do `generate_client_volfile` in csi/volumeutils.py today, generate client file, and respond.
3. Provide a mount binary/script which does a HTTP GET call internally, and fetches the volfile.


Phase2 - Dynamic

* Implement a websocket and update the client volfile, and send SIGHUP signal to mount when client volfile changes.

----
* This REST API server can be residing in kadalu operator pod itself as a container, so it is easier to watch for config change.
* Ideally, great if moana project also implements a feature like this so the mount scripts can be same for both.

**Reply by leelavg** [reply]
Looks cool, is it possible to answer 'why' such a feature is required? Is it some kubevirt use case?

**Reply by aravindavk** [reply]
It is kind of generic use case.

* Apps running outside the K8s and only Storage is provisioned from K8s
* Storage k8s Cluster is different than the apps K8s Cluster

**Reply by leelavg** [reply]
Makes sense, thanks for the explanation.

**Comment by aravindavk** [comment]
Looks good. One thing to explore is that the availability of Server pods from outside. Let us say If the generated Volfile contains server hostname as `server1.kube1` and how to make that accessible as is from the external node?

* Option 1: Update Client machines `/etc/hosts` to respective IP mappings(Difficult if the public IP of these server pods may be changing)
* Option 2: Expose these server hostnames to client machine outside the K8s cluster.

Similar thing is required for Operator pod as well. Need a way to access the ReST APIs from Operator Pod.

**Comment by aravindavk** [comment]
Another issue is related to Authentication. ReST API can implement auth roles so that only authorized will get the Volfile. But how to handle if server pods are directly accessible.

* If Full Volume mount is required - TLS/SSL based client-brick communication is the possibility
* Subvolume/PV access - Not sure how to prevent a Client from accessing another Volume other than the one exposed.
